src/rgw/driver/rados/policy.cc

```diff
@@ -0,0 +1,140 @@
+#include "rgw_zone.h"
+#include "svc_mdlog.h"
+#include "rgw_policy.h"
+#include "account.h"
+#include "cls/user/cls_user_client.h"
+
+namespace rgwrados::policy
+{
+static const std::string oid_prefix = "customer-managed-policy.";
+static constexpr std::string_view policy_oid_prefix = "policies.";
+
+void resource_metadata::dump(ceph::Formatter* f) const
+{
+  encode_json("policy_id", policy_id, f);
+}
+
+void resource_metadata::generate_test_instances(std::list<resource_metadata*>& o)
+{
+  o.push_back(new resource_metadata);
+  auto m = new resource_metadata;
+  m->policy_id = "id";
+  o.push_back(m);
+}
+
+static std::string get_name_key( const std::string_view& account_id, const std::string_view& policy_name)
+{
+  std::string lower_name(policy_name);
+  boost::algorithm::to_lower(lower_name);
+  return string_cat_reserve(oid_prefix, account_id, ".", lower_name);
+}
+rgw_raw_obj get_name_obj(const RGWZoneParams& zone, const ManagedPolicyInfo& info)
+{
+    return {zone.policy_pool, get_name_key(info.account_id, info.name)};
+}
+
+static std::string get_policy_key(std::string_view account_id) {
+  return string_cat_reserve(policy_oid_prefix, account_id);
+}
+rgw_raw_obj get_policy_obj(const RGWZoneParams& zone,
+                          std::string_view account_id) {
+  return {zone.policy_pool, get_policy_key(account_id)};
+}
+
+static int add(const DoutPrefixProvider* dpp,
+        optional_yield y,
+        librados::Rados& rados,
+        const rgw_raw_obj& obj,
+        const ManagedPolicyInfo& info,
+        bool exclusive, uint32_t limit)
+{
+  resource_metadata meta;
+  meta.policy_id = info.id;
+
+  cls_user_account_resource resource;
+  resource.name = info.name;
+  resource.path = info.path;
+  encode(meta, resource.metadata);
+
+  rgw_rados_ref ref;
+  int r = rgw_get_rados_ref(dpp, &rados, obj, &ref);
+  if (r < 0) {
+    return r;
+  }
+
+  librados::ObjectWriteOperation op;
+  ::cls_user_account_resource_add(op, resource, exclusive, limit);
+  return ref.operate(dpp, std::move(op), y);
+}
+
+static int remove_index(const DoutPrefixProvider* dpp, optional_yield y,
+                        RGWSI_SysObj& sysobj, IndexObj& index)
+{
+  int r = rgw_delete_system_obj(dpp, &sysobj, index.obj.pool,
+                                index.obj.oid, &index.objv, y);
+  if (r < 0) {
+    ldpp_dout(dpp, 20) << "WARNING: failed to remove " << index.obj << " with " << cpp_strerror(r) << dendl;
+    return r;
+  }
+  return 0;
+}
+
+static int write_path(const DoutPrefixProvider* dpp, optional_yield y,
+                      librados::Rados& rados, RGWSI_SysObj& sysobj,
+                      const RGWZoneParams& zone, const ManagedPolicyInfo& info,
+                      PathIndex& index)
+{
+  // add the new policy to its account
+  AccountIndex path;
+  path.obj = get_policy_obj(zone, info.account_id);
+  path.name = info.path;
+
+  constexpr bool exclusive = true;
+  constexpr uint32_t no_limit = std::numeric_limits<uint32_t>::max();
```

Comment:
is this creating a new object? I don't think this is correct, the idea is to add policy as a resource to the account's omap. Check code for groups, the value of exclusive is set to false
```
// link the new name to its account
    const auto& users = account::get_groups_obj(zone, info.account_id);
    r = groups::add(dpp, y, rados, users, info, false,
                    std::numeric_limits<uint32_t>::max());
```